Contributing Thales (SafeNet Trusted Access) Pack

ujji8055 opened the pull request:
## Description
The pack allows Cortex XSOAR to interact with SafeNet Trusted Access(STA) and applies security remediation actions on SafeNet Trusted Access alerts through Cortex XSOAR Security Orchestration Playbooks.

**Flow**
1. Cortex XSOAR listen to SafeNet Trusted Access Alert using Mail Listerner.
2. Cortex XSOAR creates an incident and invokes a playbook that executes the security remediation steps.
3. Playbook adds the user to a group for a configured amount of time on which STA policies with stronger authentication requirements are imposed.

**Configuration steps at Cortex XSOAR**
1. Install and configure SafeNet Trusted Access integration instance.
2. Install and configure Mail listener integration instance.
3. Add XSOAR lists for group and time duration.
4. Create a pre-process rule.

**Command**
This pack contain 17 commands.
- sta-get-user-list	:	Get list of users in the tenant.
- sta-get-user-info	:	Get profile information of a specific user.
- sta-create-user	:	Create new user in the tenant.
- sta-update-user-info	:	Update profile of a specific user.
- sta-delete-user	:	Delete user from the tenant.
- sta-get-user-groups	:	Return all the groups associated with specific user.
- sta-get-group-list	:	Get list of groups in the tenant.
- sta-get-group-info	:	Get information of a specific group.
- sta-get-group-members	:	Get list of users in a specific group.
- sta-create-group	:	Create new group in the tenant.
- sta-delete-group	:	Delete group from the tenant.
- sta-update-group-info	:	Update information of a specific group.
- sta-add-user-group	:	Add user to a specific group.
- sta-remove-user-group	:	Remove user from a specific group.
- sta-user-exist-group	:	Check if user exist in the group.
- sta-get-logs	:	Get access, authentication and audit logs.
- sta-validate-tenant	:	Checks if you have permission to access the requested tenant.

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/demisto/content?pullRequest=14175) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/demisto/content?pullRequest=14175) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/demisto/content?pullRequest=14175) it.</sub>

marcellarichmond commented:
@mchasepan

tallieber commented on Packs/SafeNet_Trusted_Access/IncidentFields/incidentfield-SafeNet_Trusted_Access_-_High_Risk_Group.json:
```diff
@@ -0,0 +1,27 @@
+{
+    "associatedToAll": true,
```
If it associate to one incident, lets write it like this (example): 
`    "associatedToAll": false,
    "associatedTypes": [
        "Infinipoint Compliant",
        "Infinipoint NotCompliant"
    ],`

ujji8055 replied:
Done for all the incident fields.

tallieber commented on Packs/SafeNet_Trusted_Access/Integrations/SafeNetTrustedAccess/SafeNetTrustedAccess.py:
```diff
@@ -0,0 +1,832 @@
+
+    response = client.get_user_groups_sta(userName=args.get('userName'), limit=args.get('limit'))
+    header_sequence = ['id', 'schemaVersionNumber', 'name', 'description', 'isSynchronized']
+    return CommandResults(
+        readable_output=tableToMarkdown(
+            f"Groups associated with user - {args.get('userName')} : ", response, headers=header_sequence,
+            headerTransform=pascalToSpace, removeNull=True),
+        outputs_prefix='STA.GROUP',
+        outputs_key_field=['id', 'name'],
+        outputs=response
+    )
+
```
In all integration commands you should handle the case the response is empty. If the response is empty we dont want to create entry on context data and just return a CommandResult with human readable.

ujji8055 replied:
Done. As discussed add below for commands.
 ```
   if not response:
        msg = 'No results found for the command.'
        return CommandResults(
            readable_output=msg,
        )
```

tallieber commented on Packs/SafeNet_Trusted_Access/Integrations/SafeNetTrustedAccess/SafeNetTrustedAccess.py:
```diff
@@ -0,0 +1,832 @@
+        query_params = {}
+        if userName:
+            self.get_user_info_sta(userName=userName)
+        if since:
+            query_params['since'] = self.validate_date_sta(datestring=since)
+        if until:
+            query_params['until'] = self.validate_date_sta(datestring=until)
```
We usually give an option to give a range "3 days" and we handle it in the code. 
Example: https://github.com/demisto/content/blob/master/Packs/FireEyeCM/Integrations/FireEyeCM/FireEyeCM.py#L219

ujji8055 replied:
Is it mandatory ?

ujji8055 replied:
As discussed, this is not mandatory. Already having filter of since, until, username.

ujji8055 commented on Packs/SafeNet_Trusted_Access/Scripts/STAPostProcessing/STAPostProcessing.yml:
```diff
@@ -0,0 +1,20 @@
+- post-processing
+- field-change-triggered
+type: python
+fromversion: 6.0.0
```
6.1.0 version installed into my machine. Not sure about 5.5.0 version.

tallieber replied:
Ok, so we will keep it from this version.

tallieber commented on Packs/SafeNet_Trusted_Access/Scripts/STAPostProcessing/STAPostProcessing.py:
```diff
@@ -0,0 +1,58 @@
+# Check if user is a member of High risk group.
+def check_user_exist_group_sta(args: Dict[str, Any]):
+    return demisto.executeCommand('sta-user-exist-group', {
+        "userName": args.get('safenettrustedaccessusername'),
+        "groupName": args.get('safenettrustedaccesshighriskgroup'),
+        "using": args.get('safenettrustedaccessinstancename'),
```
If there is arguments for the script, please add it to yml file and readme.

ujji8055 replied:
These are the fields and will rename variable with **field**.

ujji8055 replied:
Done.

tallieber commented on Packs/SafeNet_Trusted_Access/Scripts/STAFetchListContent/STAFetchListContent.py:
```diff
@@ -0,0 +1,38 @@
+# Get High Risk Group name from the list.
+def get_high_risk_group_sta(args: Dict[str, Any]) -> CommandResults:
+
+    list_name = args.get('list_name')
```
If this is a script argument, please add to yml and readme.

ujji8055 replied:
As discussed, no change needed.

ujji8055 commented on Packs/SafeNet_Trusted_Access/pack_metadata.json:
```diff
@@ -0,0 +1,26 @@
+    "currentVersion": "1.0.0",
+    "author": "Thales",
+    "url": "https://supportportal.gemalto.com/csm/?id=portal_home_page",
+    "email": "",
```
Will share the same later.

ujji8055 replied:
Email : technical.support.DIS@thalesgroup.com 
However, this is not for opening tickets. Tickets have to be created / entered directly on the support portal.

ujji8055 commented:
@tallieber  I am facing merge conflict issue. Can you please check and suggest the needful.

ujji8055 commented:
@evisochek  @tallieber 
1) As discussed, updated the term "High Risk Group" to "Unusual Activity Group" in all the components.
2) Added first task **IsIntegrationAvailable** in the playbook and added condition as suggested.
3) Connected **Failed message** tasks to the Done section as we discussed.
4) Mapped the mapper with required fields that we'll show in our next demo.
5) Uploaded updated image after your modification.
6) Created readme file of Playbook and linked to the updated image.
7) Removed unnecessary fields from your layout JSON file.

ujji8055 commented on Packs/SafeNet_Trusted_Access/Classifiers/classifier-SafeNet_Trusted_Access_–_Push_Reject_Alert_Classifier.json:
```diff
@@ -0,0 +1,68 @@
+{
+    "description": "Classifies SafeNet Trusted Access alerts.",
+    "feed": false,
+    "id": "b3e49867-6677-42eb-808d-5ba1b25d5e6d",
```
Done

ujji8055 commented on Packs/SafeNet_Trusted_Access/Integrations/SafeNetTrustedAccess/SafeNetTrustedAccess.py:
```diff
@@ -0,0 +1,896 @@
+        readable_output=tableToMarkdown("List of users in the tenant :", response, headers=header_sequence,
+                                        headerTransform=pascalToSpace, removeNull=True),
+        outputs_prefix='STA.USER',
+        outputs_key_field=['id', 'userName'],
```
Done. As discussed removed **outputs_key_fields** value from unnecessary places.

ujji8055 commented on Packs/SafeNet_Trusted_Access/Layouts/layoutscontainer-SafeNet_Trusted_Access_-_Incident_Layout.json:
```diff
@@ -0,0 +1,1936 @@
+                    },
+                    {
+                        "fieldId": "incident_servicenowticketnumber",
+                        "isVisible": true
```
Done. Removed servicenow fields from layout.

ujji8055 commented on Packs/SafeNet_Trusted_Access/Playbooks/SafeNet_Trusted_Access_–_Add_to_High_Risk_Group.yml:
```diff
@@ -0,0 +1,842 @@
+  }
+fromversion: 6.0.0
+tests:
+- No tests (auto formatted)
```
Done. Added test playbook (Test Playbook STA) to the playbook.

tallieber commented on Tests/conf.json:
```diff
@@ -3857,4 +3857,4 @@
     }
-}
\ No newline at end of file
+}
```
this is by mistake, right?

ujji8055 replied:
Yes. Earlier, I was experiencing conflict in this file while merging the code. As suggested by Kobbi, I copied the content from master repo to my local machine and then pushed the code and it works fine.

tallieber commented on Packs/SafeNet_Trusted_Access/Integrations/SafeNetTrustedAccess/SafeNetTrustedAccess.py:
```diff
@@ -0,0 +1,935 @@
+    client.validate_tenant_sta()
+    return CommandResults(
+        readable_output="## The requested tenant is accessible.",
+        outputs_prefix='STA.VALIDATE.TENANT',
+        outputs=True
```
If you dont need this data for further use, so I dont need to create a context data, you can just return a human readable

ujji8055 replied:
Done. Removed outputs and outputs_prefix field from **sta-validate-tenant** command.

ujji8055 replied:
@tallieber 
We are using this context data in the playbook. So no need to remove the same.

ujji8055 commented:
@tallieber
Please suggest the next step for the release process.

tallieber commented on Packs/SafeNet_Trusted_Access/Integrations/SafeNetTrustedAccess/SafeNetTrustedAccess.py:
```diff
@@ -0,0 +1,935 @@
+        readable_output=tableToMarkdown(
+            f"Groups associated with user - {args.get('userName')} : ", response, headers=header_sequence,
+            headerTransform=pascalToSpace, removeNull=True),
+        outputs_prefix='STA.GROUP',
+        outputs_key_field=['id'],
+        outputs=response
```
If this command return all group the user is associated to, so where we can see it in the context? You can list all users under a group, or list all user's groups under USER

ujji8055 replied:
Done.
As suggested, updated the context data for **user-groups** and **group-members** commands.
![image](https://user-images.githubusercontent.com/86044314/132178794-9d89591d-7645-4967-8597-de76929ab9d7.png)
![image](https://user-images.githubusercontent.com/86044314/132178892-43f7565f-02d7-467e-aabe-7bc8ee40e933.png)

ujji8055 commented:
@tallieber 
I have incorporated the changes. Please let me know if anything else is required from our end. 
Can you please suggest the next step to get the content pack into the marketplace?
